File: LeetCode/0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates, target):
        def dfs(path, sum_path):
            if sum_path == target:
                sorted_path = path[:]
                sorted_path.sort()
                if sorted_path not in answer:
                    answer.append(sorted_path)
                return

            if sum_path > target:
                return

            for i in range(len(candidates)):
                if (sum_path + candidates[i] <= target):
                    path.append(candidates[i])
                    dfs(path, sum_path + candidates[i])
                    path.pop()

        answer = []
        dfs([], 0)
        return list(answer)

File: LeetCode/0039-combination-sum/0039-combination-sum.java
class Solution {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> answer = new ArrayList<>();  
        dfs(0, 0, new ArrayList<>(), candidates, target, answer);
        return answer;
    }

    private void dfs(int depth, int sum, List<Integer> path, int[] candidates, int target, List<List<Integer>> answer) {
        if (sum > target) {
            return;
        }
        if(sum == target){
            answer.add(new ArrayList<>(path));
            return;
        }

        for(int i = depth; i < candidates.length; i++) {
            path.add(candidates[i]);
            dfs(i, sum + candidates[i], path, candidates, target, answer); 
            path.remove(path.size() - 1);  
        }
    }
}